tokio/src/runtime/task/mod.rs

```diff
@@ -71,10 +72,33 @@
 //!    a lock for the stage field, and it can be accessed only by the thread
 //!    that set RUNNING to one.
 //!
-//!  * If JOIN_WAKER is zero, then the JoinHandle has exclusive access to the
-//!    join handle waker. If JOIN_WAKER and COMPLETE are both one, then the
-//!    thread that set COMPLETE to one has exclusive access to the join handle
-//!    waker.
+//!  * The waker field may be concurrently accessed by different threads: in one
+//!    thread the runtime may complete a future and read the waker field to
+//!    invoke the waker, and in another thread the future's JoinHandle may be
+//!    polled and if the future hasn't completed, JoinHandle may install a waker
+//!    in the waker field. The JOIN_WAKER bit ensures safe access by multiple
+//!    threads to the waker field using the following rules:
+//!
+//!    1. JOIN_WAKER is initialized to zero.
+//!
+//!    2. If JOIN_WAKER is zero, then the JoinHandle has exclusive (mutable)
+//!       access to the waker field.
+//!
+//!    3. If JOIN_WAKER is one, then the JoinHandle has shared (read-only)
+//!       access to the waker field.
+//!
+//!    4. If the JoinHandle needs to install a waker, then the JoinHandle needs
+//!       to (i) successfully set JOIN_WAKER to zero if it is not already zero
+//!       to gain exclusive access to the waker field per rule 2, (ii) install a
+//!       waker, and (iii) successfully set JOIN_WAKER to one.
+//!
+//!    5. The JoinHandle can only change JOIN_WAKER if COMPLETE is zero (i.e.
+//!       the task hasn't yet completed). Hence, steps (i) or (iii) of rule 4
+//!       may fail due to a race.
+//!
+//!    6. Once COMPLETE is one (i.e. task has completed), the JoinHandle will
+//!       not modify JOIN_WAKER and the waker field. At this point the runtime
+//!       invokes the waker if there is one.
```

Comment:
The join handle _can_ still modify the waker field when COMPLETE is one. This happens if COMPLETE is changed to one during step 4. (ii).